File: rankWithTies/rankWithTies.m
% rankWithTies - Ranking with tie-break control
%{ 
%-------------------------------------------------------------------------------
% SYNTAX:
%   [R,B,I] = rankWithTies(A,method,<sortOpts>)
%
% PURPOSE:
%   This function performs ranking with additional control of tie-breaks.  
%   Ranking is fundamentally a sorting operator where one reports the relative
%   position of each element within the sorted list.  When ties are present,
%   these relative positions become ambiguous, and this function controls how 
%   those positions get reported.
% 
%   The outputs R and I are indices that describe permutation operators with the
%   properties A = B(R) and B = A(I) for vector-valued inputs.  R and I will not
%   represent inverse permutation operators when ties are present if one does
%   not ensure unique(R) = unique(I).  For example, the 'min' and 'max' methods
%   below will lead to R and I representing permutations that are not inverses
%   of one another when ties are present.
%  
% INPUT:
%   A           - Array to be sorted
%
%   method      - Method for breaking ties.  Common defaults are provided and 
%                 one can also define an arbitrary function that acts on the
%                 stable ranks
%       'min'       - Smallest index               
%       'max'       - Largest index
%       'random'    - Random permutation index
%       'stable'    - Order of occurance in A
%
%                 A custom function defined on the stable ranks given by
%                 (rMax-nMatch):rMax.  This function must have to prototype
%                   ROut(ind) = methodFun(rMax,nMatch)
%                 where rMax is the maximum stable rank and nMatch is the
%                 number of matches within the set B( (rMax-nMatch):rMax ).
%                 For the aforementioned defaults this function is:
%       'min'       - @(rMax,nMatch) rMax-nMatch
%       'max'       - @(rMax,nMatch) rMax
%       'random'    - @(rMax,nMatch) randperm(nMatch+1) + (rMax-nMatch-1)
%       'stable'    - @(rMax,nMatch) (rMax-nMatch):rMax 
%                 
%
%   sortOpts    - Options to built-in sort call
% 
% OUTPUT:
%   R       - Ranks.  Reverse sort index s.t. A = B(R)
%   B       - Sort result
%   I       - Sort index with ties broken via method s.t. B = A(I)
%
% ASSUMPTIONS: 
%   All input variables are of the correct type, valid (if applicable),
%   and given in the correct order.
%
% NOTES:
%   Vectorization doesn't explicitly parallelize replaceDuplicateRanks 
%
% SEE ALSO:
%   datafun/sort
%-------------------------------------------------------------------------------
%}
function [R,B,I] = rankWithTies(A,method,varargin)

% Interpret cell array sortOpts as a cell of options.  This nicety works because
% sort does not accept options of type cell.
if nargin==3 && iscell(varargin{1})
    varargin = varargin{1};
end

% Determine if vectorization is being used
vecDim = getVectorizationDim(A,varargin{:});

% Sort
[B,I] = sort(A,varargin{:});

% Get the stable ranks (reverse indices)
R = getStableRanks(I,vecDim);

% Get the duplicate rank function
switch class(method)
    case 'char'
        % Provide the defaults
        switch lower(method)
            case 'min'
                % min( R(I( (rMax-nMatch):rMax )) )
                fun = @(rMax,nMatch) rMax-nMatch;
                
            case 'max'
                % max( R(I( (rMax-nMatch):rMax )) )
                fun = @(rMax,nMatch) rMax;
                
            case {'random','rand'}
                % Random permutation of R(I( (rMax-nMatch):rMax  ))
                fun = @(rMax,nMatch) randperm(nMatch+1) + (rMax-nMatch-1);
                
            case 'stable'
                % Nothing left to do
                return
                
            otherwise
                error([mfilename(),':unknownMethod'],...
                    'Unknown tie-break method: %s', ...
                    strtrim(evalc('disp(method)')) );
        end
        
    case 'function_handle'
        % User provided function
        fun = method;
        
    otherwise
        error([mfilename(),':invalidMethodType'],...
                    'Invalid method type: %s', class(method) );
end

% Replace duplicate stable ranks according to fun
if vecDim
    R = replaceDuplicateRanksVec(B, I, R, fun, vecDim);
else
    R = replaceDuplicateRanks(B, I, R, fun );
end


end


%{ 
%-------------------------------------------------------------------------------
% SYNTAX:
%   vecDim = getVectorizationDim(A,varargin)
%
% PURPOSE:
%   Returns the working dimension if vectorization is being used.
%  
% INPUT:
%   See main function
% 
% OUTPUT:
%   vecDim      - Vectorization dimension or [] if no vectorization
%
% ASSUMPTIONS: 
%   All input variables are of the correct type, valid (if applicable),
%   and given in the correct order.
%-------------------------------------------------------------------------------
%}
function vecDim = getVectorizationDim(A,varargin)

% Default Values
ASize = size(A);
    
if numel(varargin) && isnumeric(varargin{1})
    % User provided an explicit dimension to sort
    vecDim = varargin{1};
    
    % Check for vectorization
    if vecDim <= numel(ASize)
        ASize(vecDim) = [];
    end
    if prod(ASize)==1
        vecDim = [];
    end
    
elseif sum(ASize>1)>1
    vecDim = 1;
    
else
    vecDim = [];
end

end


%{ 
%-------------------------------------------------------------------------------
% SYNTAX:
%   R = getStableRanks(I,vecDim)
%
% PURPOSE:
%   Returns the stable ranks associated with I.  This is the reverse permutation
%   indices s.t. R(I) are ranks of elements that are sorted by I.
%  
% INPUT:
%   I           - Sort index s.t. B = A(I) sorts A
%   vecDim      - Vectorization dimension or [] if no vectorization
% 
% OUTPUT:
%   R           - Stable ranks s.t. A = B(R)
%   IFull       - I described in terms of the full vectorized problem
%
% ASSUMPTIONS: 
%   All input variables are of the correct type, valid (if applicable),
%   and given in the correct order.
%-------------------------------------------------------------------------------
%}
function [R,I] = getStableRanks(I,vecDim)

% Default Values
ISize = size(I);

if isempty(vecDim)
    % Get stable reverse indices
    N = numel(I);
    R(I) = 1:N;
    
    % Ensure R is of the correct size
    if ISize(2)~=N || ISize(2)==0
        R = reshape(R,ISize);
    end
    
else    
    % Trivial vectorization
    if vecDim > numel(ISize)
        R = I;
        return
    end
    
    % Get the vectorized assignments "1:N"
    nDim = numel(ISize);
    for i = 1:nDim
        if i==vecDim
            ind{i} = ':';
        else
            ind{i} = ones(ISize(i),1);
        end
    end
    RSize = ones(1,nDim);
    RSize(vecDim) = ISize(vecDim);
    R = reshape(1:ISize(vecDim), RSize);
    R = R(ind{:});
    
    % Compute the delta along vecDim
    delta = prod(ISize(1:vecDim-1));
    
    % Convert I to the linear index associated with the full problem
    I = reshape(1:prod(ISize),ISize) + (I-R) * delta;
    
    % Transform R into the desired reverse mapping
    R(I) = R;
    
end

end


%{ 
%-------------------------------------------------------------------------------
% SYNTAX:
%   R = replaceDuplicateRanks(B,I,R,fun)
%
% PURPOSE:
%   Replace duplicate ranks with a function of their stable ordering.
%  
% INPUT:
%   B       - Sorted output
%   I       - Forward sort index s.t. B = A(I)
%   R       - Stable reverse sort index s.t. A = B(R)
%   fun     - Function mapping stable ranks to their desired output where the
%             stable ranks are given by (rMax-nMatch):rMax
%               ROut(ind) = fun(rMax,nMatch)
% 
% OUTPUT:
%   R       - Updated reverse sort index (ranks) s.t. A = B(R)
%
% ASSUMPTIONS: 
%   All input variables are of the correct type, valid (if applicable),
%   and given in the correct order.
%-------------------------------------------------------------------------------
%}
function R = replaceDuplicateRanks(B,I,R,fun)

% Default Values
N = numel(B);
matchCount = 0;

for i = 1:N-1
    if B(i)==B(i+1)
        matchCount = matchCount + 1;
        
    elseif matchCount
        % Resolve the matched indices
        % ind = (i-matchCount):i;
        % R(I(ind)) = fun( R(I(ind)) );
        R(I( (i-matchCount):i )) = fun(i,matchCount);
        
        matchCount = 0;
    end

end

% Resolve remaining matches
if matchCount
    R(I( (N-matchCount):N )) = fun(N,matchCount);
end

end



%{ 
%-------------------------------------------------------------------------------
% SYNTAX:
%   R = replaceDuplicateRanksVec(B,I,R,fun,vecDim)
%
% PURPOSE:
%   Vectorize replaceDuplicateRanks.  The most efficient path is not yet clear.
%  
% INPUT:
%   B       - Sorted output
%   I       - Forward sort index s.t. B = A(I)
%   R       - Stable reverse sort index s.t. A = B(R)
%   fun     - Function mapping stable ranks to their desired output where the
%             stable ranks are given by (rMax-nMatch):rMax
%               ROut(ind) = fun(rMax,nMatch)
%   vecDim  - Dimension of vectorization
%
% OUTPUT:
%   R       - Updated reverse sort index (ranks) s.t. A = B(R)
%
% ASSUMPTIONS: 
%   All input variables are of the correct type, valid (if applicable),
%   and given in the correct order.
%-------------------------------------------------------------------------------
%}
function R = replaceDuplicateRanksVec(B,I,R,fun,vecDim)

% Default Values
BSize = size(B);
nDim = numel(BSize);

if numel(BSize)<vecDim
    % Trivial vectorization
    R = replaceDuplicateRanks(B,I,R,fun);
    return
elseif numel(B) == 0
    % Trivial vectorization
    R = zeros(BSize);
    return
end

% Place the working dimension up front
if prod(BSize(1:vecDim-1)) > 1
    order = 0:nDim;
    order(vecDim+1) = [];
    order(1) = vecDim;
    
    B = permute(B,order);
    I = permute(I,order);
    R = permute(R,order);
else
    order = [];
end

% Iterate over each slice
N = numel(B)/BSize(vecDim);
for i = 1:N
    R(:,i) = replaceDuplicateRanks(B(:,i),I(:,i),R(:,i),fun);
end

% Inverse permute
if order
    R = ipermute(R,order);
end

end

File: rankWithTies/rankWithTies.py
import numpy as np

def rankWithTies(A, method, *args):
    '''
    rankWithTies - Ranking with tie-break control
     SYNTAX:
       R,B,I = rankWithTies(A,method,<sortOpts>)

    PURPOSE:
    This function performs ranking with additional control of tie-breaks.
    Ranking is fundamentally a sorting operator where one reports the relative
    position of each element within the sorted list.  When ties are present,
    these relative positions become ambiguous, and this function controls how
    those positions get reported.

    The outputs R and I are indices that describe permutation operators with the
    properties A = B[R] and B = A[I] for vector-valued inputs.  R and I will not
    represent inverse permutation operators when ties are present if one does
    not ensure unique[R] = unique[I].  For example, the 'min' and 'max' methods
    below will lead to R and I representing permutations that are not inverses
    of one another when ties are present.

     INPUT:
       A           - Array to be sorted

       method      - Method for breaking ties.  Common defaults are provided and
                     one can also define an arbitrary function that acts on the
                     stable ranks
           'min'       - Smallest index
           'max'       - Largest index
           'random'    - Random permutation index
           'stable'    - Order of occurance in A

                     A custom function defined on the stable ranks given by
                       lambda rMax, nMatch :
                     This function must have to prototype
                       ROut(ind) = methodFun(rMax,nMatch)
                     where rMax is the maximum stable rank and nMatch is the
                     number of matches within the set B[rMax-(nMatch-1):rMax].
                     For the aforementioned defaults this function is:
           'min'       - lambda rMax, nMatch : rMax-(nMatch-1)
           'max'       - lambda rMax, nMatch : rMax
           'random'    - lambda rMax, nMatch :
                        np.random.permutation(nMatch) + (rMax - (nMatch - 1))
           'stable'    - lambda rMax, nMatch:np.arange((rMax-(nMatch-1)),rMax+1)


       sortOpts    - Options to numpy built-in sort call

     OUTPUT:
       R       - Ranks.  Reverse sort index s.t. A = B[R] along dim
       B       - Sort result
       I       - Sort index with ties broken via method s.t. B = A[I] along dim

     ASSUMPTIONS:
       All input variables are of the correct type, valid (if applicable),
       and given in the correct order.

     NOTES:
       Vectorization doesn't explicitly parallelize replaceDuplicateRanks
    '''
    #stores calls to built in methods
    DEFAULT_METHODS = {'min': lambda rMax, nMatch : rMax - (nMatch-1),
                       'max': lambda rMax, nMatch : rMax,
                       'random': lambda rMax, nMatch : \
                        np.random.permutation(nMatch) + (rMax - (nMatch - 1))
                      }

    #if no dimension specified, assume ranking along last dimension of A
    if not len(args):
        args += (-1,)

    dim = args[0]

    I = np.argsort(A, *args)
    #avoids sorting twice
    #indexes A along 1D slices in dim using indices in respective I slices
    B = np.take_along_axis(A, I, dim)

    #initialization of inverse index array
    R = np.zeros_like(I)

    N = A.shape[dim]

    #reshapes the arange(N) array to be along the axis to insert in R
    indexer = tuple([None if i!=dim else Ellipsis for i in range(A.ndim)])

    np.put_along_axis(R, I, np.arange(N, dtype = float)[indexer], axis=dim)

    if method == 'stable':
        return R, B, I

    elif method in DEFAULT_METHODS.keys():
        fun = DEFAULT_METHODS[method]

    elif callable(method):
        fun = method

    else:
        raise ValueError(f"Invalid method type: {method}")

    #move axis to loop through to end for ease of iteration/ dim generalization
    B_shift = np.moveaxis(B, dim, -1)
    I_shift = np.moveaxis(I, dim, -1)
    R_shift = np.moveaxis(R, dim, -1)

    for index in np.ndindex(B_shift.shape[:-1]):
        #perform rank tie-breaking along 1D slices
        R_shift[index] = replaceDuplicateRanks(B_shift[index], \
                                        I_shift[index], R_shift[index], fun)

    return R, B, I

def replaceDuplicateRanks(B, I, R, fun):

    #arrays with the indices the unique values start/stop at and their distances
    uniq, uniq_ind, uniq_count = \
                        np.unique(B, return_index = True, return_counts = True)

    #makes sure the to check for the last value being a duplicate
    uniq_ind = np.append(uniq_ind, len(B))

    for i in range(len(uniq_ind)-1):
        if uniq_count[i] != 1:
            #replaces rank ties according to fun
            R[I[uniq_ind[i]:uniq_ind[i+1]]]= fun(uniq_ind[i+1]-1, uniq_count[i])

    return R
